### DISCUSSION TITLE
A question about the implementation of B-bar in StressDivergenceTensors

### DISCUSSION THREAD

**Discussion by TJT-post95** [idaholab/moose#25965] [Q&A Modules: Solid mechanics] [answered]
Hello MOOSE:

In the simulation of incompressible material, the B-bar method is applied in MOOSE to solve the volumetric locking problem. While reading the code in `StressDivergenceTensors`, I am temporarily stuck with confusion. For example, in the member function `computeQpJacobian()`, only B0*(C0000+C0011+C0022)*(Bvol_0-B0)/3 seems to be computed for B^T_i * C * Bvol_j, where for x direction (`_component` = 0). 

According to Bower's Applied Mechanics of Solids, it seems that (B0*(C0000+C0011+C0022)+B1*(C0100+C0111+C0122)+B2*(C0200+C0211+C0222))*(Bvol_0-B0)/3 would be here?

I'm looking forward to the comments.

Thanks!

**Comment by GiudGiud** [comment]
@cpritam @sveerara

**Comment by TJT-post95** [comment]
I will take the Cook's Membrane mentioned by MOOSE as a benchmark to test the implementation of B-bar in `StressDivergenceTensors`.

**Reply by GiudGiud** [reply]
@recuero

**Reply by TJT-post95** [reply]
Thanks！😄

**Comment by recuero** [comment, marked as answer]
It seems that your analysis is focused on the Jacobian calculations, which do not change your simulation results (they can, for sure, change its convergence behavior). You can use the finite difference preconditioner to check the correctness of the Jacobian calculations for the B-bar formulation and compare with your own.

**Reply by TJT-post95** [reply]
OK！I will have a try.
